[AN-Issue-1331] Supra Native Automation Feature implemenation

aregng opened the pull request.

nizam-supraoracles committed:
feat - initial commite for supra automation registry smart contract

nizam-supraoracles committed:
restructe automation registry contract

Commit:
[AN-Issue-1337] Added/updated types to facilitate automation task handling and execution

- Added a new transaction type AutomatedTransaction for automation-tasks
- Updated TransactionMetadata to include hash value calculated from transaction bytes
- Updated UserTransactionContext to include transaction hash from bytes
- Added new native function to retrieve transaction-hash from bytes from UserTransactionContext

Commit:
Enabled ci for release and feature branches and PRs on them

Commit:
Updated api types to support automated transactions

nizam-supraoracles committed:
registry contract refactor functions and the struct

nizam-supraoracles committed:
chore - registry sc, commenting correction

nizam-supraoracles committed:
introduce supra-lib enumerable_map

Commit:
Addressed review comments

nizam-supraoracles committed:
supra automation - replace table with enumerable_map and did related updates

Commit:
Fixed copyright update

nizam-supraoracles committed:
move enumerable_map file into supra-stdlib, add `gas_committed_for_next_epoch` field, add more assersion in `register` function, also `get_active_task_ids` function will return only active task ids

Commit:
Fixec copywrite and *.md File source

- Updated move files to avoid *.md files regeneration
- Updated some *.md files to replicate the latest version of the sources

nizam-supraoracles committed:
automation registry - implement `on_new_epoch`, `update_automation_gas_limit` and `update_duration_upper_limit`

nizam-supraoracles committed:
registry function modify as get tx_hash from `txn_app_hash` and `gas_price_cap` should not be zero also collect static registry fee from user

nizam-supraoracles committed:
add `test_registry` unit testcase in supra automation registry

nizam-supraoracles committed:
automation_registry - implement `withdraw_fee` and `remove_task`, also include missing events

nizam-supraoracles committed:
address PR comments

nizam-supraoracles committed:
create `refund_automation_task_fee` function and common function `transfer_fee_to_account_internal`

nizam-supraoracles committed:
create view function `get_gas_committed_for_next_epoch`

aregng committed:
[AN-Issue-1346] Implemented custom handling of automation transaction (#134)

* [AN-Issue-1346] Implemented custom handling of automation transactions

- Added separate flow for automation transaction handling. It differes
from user transaction flow only by custom validation step of
inner-payload of automation transaction

- Added validation logic for the automation transaction inner-payload

- Introuced specific StatusCode for failures during automation
transaction inner-payload validation. They are qualified as verification
status type and will lead to have transaction kept and the account will
be charged.

* Updated copyright info

* Added language-e2e-testsuite to ci-flow with fixed tests

* Addressed review comments and added tests

- Introduced a new type PayloadTypeReference which will describe
  user-transaction payload type and reference payload if required.

- Updated TransactionMetadata and UserTransctionContext to utilize
  PayloadTypeReference enabling easy extension of any new payload type
  when needed

- Updated
  transaction_arg_validation::validate_combine_signer_and_txn_args to
  accept GasMeter as input argument.

- Updated automation inner payload validation to charge gas for it

- Added initial set of e2e tests to check Automation transaction flow.

* Fixed test failures after rebase

* Fixed copyright comment

* Introduce means to create and process automation transaction specifying the required arguments for registration

* Addressed review comments

* Renamed 'automation transaction' to 'automation registration'

* Updated automation::registry to be non-entry and made direct call from native layer

* Cosmetic chagnes

* Copyright fix

* Temporarly fixed test failure

* Split and reordered test runs in ci

* Increased nextest timeout from 2100 to 3000

---------

Co-authored-by: Aregnaz Harutyunyan <>

Commit:
[AN-Issue-1394] Added AptosVMViewer wrapper on AptosVM for continous view function execution

- AptosVMViewer provides means to continously execute view functions on
the same state-view without recreating VM over and over again compared
to AptosVM. It provides better about 7-8 times better preformance
compared to AptosVM. This will help to retrieve automation
task inforamtion in more optimal way.

- Added Rust variant of AutomationTaskMetaData and provided means to
build AutomatedTransactions based on AutomationTaskMetaData

- Enabled aptos-types unit-tests in CI flow. Ignored for the time being
the tests which are currently failing.

- Added tests for newly introduced APIs/fnctionalities

Commit:
Fixed a comment

aregng committed:
[AN-Issue-1409] Handled automation-registry state update on new epoch (#144)

* Handled automation-registry state update on new epoch

- split automation registry external API and internal state to avoid circular dependencies
- enabled automation-registry-state update on new epoch from block module.

* Addressed review comments

* Added couple of unit-tests for automation-registry-state

* Updated automation-registry to accept registration transaction hash as argument

- Moved basic parameters checks of registration to automation_registry_state
- Added and updated unit tests of automation_registry_state

* Removed redundant imports.

* Addressed review comments

---------

Co-authored-by: Aregnaz Harutyunyan <>

nizam-supraoracles committed:
enumerableMap `update_value` function is return old value

nizam-supraoracles committed:
enumerableMap include `for_each_value` and `for_each_value_ref` method

nizam-supraoracles committed:
enumerableMap - create 'filter' method

nizam-supraoracles committed:
enumerableMap PR comments address

Commit:
Fixed transaction parent hash calculation to match txn hash in smr-moonshot

- smr moonshot uses Keccak::256 to calcualte the hash of the data.
  Introduced `HashValue::keccak_256_of` API and utilized it to calcuate
  `TransactionMetadata::txn_app_hash`

nizam-supraoracles committed:
enumerable_map - implement `map` `map_ref` and `filter_map` methods and there unit tests

nizam-supraoracles committed:
enumerable_map - access properties via function instead of  direct access to properties

nizam-supraoracles committed:
automation registry - `get_active_task_ids` function is now usitlise enumerable_map's `filter_map` method

aregng committed:
[AN-Issue-1347] Implemented Automated transaction execution flow (#136)

* [AN-Issue-1347] Implemented Automated transasction execution flow

* Updated automated task prologue and added e2e tests

- Added chain-id verification and task availability check by task index
- Added e2e tests for AutoamtedTransaction

* Fixed test failure

* Added task sender check in automated_transaction_prologue

---------

Co-authored-by: Aregnaz Harutyunyan <>

aregng committed:
[AN-Issue-1432] Convenience updates to facilitate supra-layer on automation tasks execution (#154)

Co-authored-by: Aregnaz Harutyunyan <>

nizam-supraoracles committed:
update `epoch_interval` and `last_reconfiguration_time` in the automation registry module

nizam-supraoracles committed:
automation registry - move regisry state all the testcases in to single module and then remove registry state module

nizam-supraoracles committed:
supra-framework - auto generated files update as per new changes in code

nizam-supraoracles committed:
registry on_new_epoch trigger from reconfigration module instead of block

nizam-supraoracles committed:
automation registry contract - addressing PR comments

nizam-supraoracles committed:
automation registry - remove duplciate `gas_committed_for_next_epoch` field + remove `last_reconfiguration_time` from on_new_epoch + remove `get_last_epoch_time_second` unused function + update testcases

nizam-supraoracles committed:
automation registry - combine `AutomationRegistryState` fields with `AutomationRegistry` and make related changes

nizam-supraoracles committed:
automation registy - introduce new `EpochState` struct to store related to epoch data

nizam-supraoracles committed:
automation registry - use of `expected_epoch_duration` field

nizam-supraoracles committed:
automation regisry pr comments address, `EpochState` rename to `AutomationEpochInfo` + remove `on_new_epoch` call from `emit_genesis_reconfiguration_event` fun + check overflow value gas_committed_for_next_epoch

nizam-supraoracles committed:
make it `DEFAULT_DURATION_UPPER_LIMIT` to `2_626_560` which is 30.4 days

nizam-supraoracles committed:
remove unused borrow mut from `cancle_task` function

nizam-supraoracles committed:
automation registry - introduce new `AutomationRegistryConfig` struct to handle configrable data

nizam-supraoracles committed:
check underflow for `gas_committed_for_next_epoch` before cancalling task

nizam-supraoracles committed:
automation registry - use `config_buffer` to store `AutomationRegistryConfig` and replace the data after on_new_epoch

nizam-supraoracles committed:
registry sc - intialize function use values from params instead using default value, make relavent test cases update

nizam-supraoracles committed:
automation registry - remove entry from update_config function + create `check_registration_task_duration` to check necessory time related operation

nizam-supraoracles committed:
automation registry - create `initializate_by_default` function until we have initialization flow properly implemented

nizam-supraoracles committed:
automation registry - make it all the structure as `resource_group_member`

nizam-supraoracles committed:
automation registry - add new configration parameters and rename old as well

nizam-supraoracles committed:
automation registry - combining, cancellation/flat registration

nizam-supraoracles committed:
automation registry - create seprate config of active one

nizam-supraoracles committed:
automation registry - rename `RefundFeeUser` to `AutomationCancellationRefund` and add task_index + create common `config_update_from_buffer` internal function + add expirty_time>current_time condition in to `refund_automation_task_fee` fun

nizam-supraoracles committed:
fix - asserstion mistake fix which included in last commit and update function name

aregng committed:
[AN-Issue-1432] Updated AutomatedTransactionBuilder with expiration time threshold value (#162)

* [AN-Issue-1432] Updated AutomatedTransactionBuilder with expiration time threshold value

* Introduced getter interfaces for AutomationTaskMetaData properties

* Updated autoamted txn builder and automation task meta data

- Added getter interface for AutomatedTransactionBuilder properties
- Updated AutomationTaskMetaData rust struct to match the one in Move
  - Updated e2e-tests to detect incorrespondance in the same types in move and rust layer

* Fixed test failure

---------

Co-authored-by: Aregnaz Harutyunyan <>

nizam-supraoracles committed:
automation registry - `refund_automation_task_fee` only if `automation_task_metadata.expiry_time > current_time`, otherwise do not fail the tx

nizam-supraoracles committed:
automation registry - autoupdate aptos_framework_sdk_builder

aregng committed:
[AN-Issue-1419] Updated Automation Registration Parameters (#170)

* [AN-Issue-1419] Updated Automation Registration Parameters

- Added 2 new properties
  - automation_fee_cap_for_epoch to allow user decide on threshold for the automation fee paid for epoch
  - aux_data - means to allow future minor registration parameters integration requiring only framework/state upgrade
- Refactor RegistrationParams to allow easy extend with new/other set of parameter lists in future

* Addressed review comments

* Fixed config existence check in 'automation-registry::update_epoch_interval_in_registry`

* Fixed the typo

---------

Co-authored-by: Aregnaz Harutyunyan <>

isaacdoidge commented on api/types/src/transaction.rs:
```diff
@@ -301,6 +310,37 @@ impl
+            u64,
+        ),
+    ) -> Self {
+        // If at some point Aptos will utilize Automated transactions then here we will use as
```
This comment isn't clear to me. How would Aptos integrate our changes? And what would change if they did?

aregng replied:
This is a public API of aptos REST interface. We are not utilizing it, but as long as we introduced a new transaction type, it should be update API types as well to keep code-base consistent and compile error free.

`AutomatedTranscation` has reference to the automation-registration-transaction hash. 
Transaction hash calculated by APOTS differs from transaction has that smr-moonshot(supra) calculates for transactions.
In current implementation of `AutomationRegistration` transcation execution in MoveVM we are using transaction-hash in supra-style to specify parent-hash of the automation task in order for later to use it as `Authenticator` in supra-layer.
Here  to create meta should have been specified registration transaction hash in Aptos Style, but it is not possible.
I think it would be better to specify registration transaction hash from `AutomatedTranscation::txn_hash` and add a comment that in the execution flow `AutomationRegistration` transaction hash calculation should be updated to APTOS style if APTOS DB and API flow is going to be utilized to request transaction data.

And I am not sure whether Automation feature is going to be used by aptos at all.

isaacdoidge replied:
Can you please add the clarification that you've provided in this comment about the nature of the hash, into the comment in the source code? That will help to make it clearer. Updating the API types to avoid compile errors makes sense. We don't have any plans to utilize Aptos' database and API logic though (aside from the API types and Converter that we're already using).

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -875,6 +883,150 @@ impl AptosVM {
+        new_published_modules_loaded: &mut bool,
+        change_set_configs: &ChangeSetConfigs,
+    ) -> Result<(VMStatus, VMOutput), VMStatus> {
+        fail_point!("aptos_vm::execute_automation_registration_txn", |_| {
```
Could you please briefly explain what this function does and how it adds value to the code? The error codes seem very general.

aregng replied:
`fail_point` helps to simulate failures if needed to check the flow with the failure in particular places.
I have inherited this code form the counterpart function executing script or entry-function.
Currently we are not doing any testing by simulating failures, so if it is redundant then I can simply remove it from here.

isaacdoidge replied:
Not sure I fully understand. Is this function intended to provide information that can be used by tests to simulate errors at certain points in the flow of execution?

aregng replied:
I think this doc will explain better than me: https://docs.rs/fail/latest/fail/index.html

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -875,6 +883,150 @@ impl AptosVM {
+        let actual_args = inner_entry_function.args().to_vec();
+        // By constructing args we are making sure that function execution in scope of automated
+        // task will not fail due to invalid arguments passed
+        verifier::transaction_arg_validation::validate_combine_signer_and_txn_args(
```
If I understand correctly, then this is why `validate_combine_signer_and_txn_args` now takes a gas meter. Do we need to charge gas when verifying the entry function of a registration transaction though? We don't appear to do this for normal transactions.

aregng replied:
for actual transaction payload validation we are not charging, but we decided to charge for inner payload validation.

isaacdoidge replied:
What is the reasoning behind this? It seems to me like this complicates the code and makes the UX inconsistent

isaacdoidge replied:
Having read your comment on my other thread about re-validation for `AutomatedTransaction`s, I see why we might have been inclined to charge for validating the inner payload. However, we should not need to validate the payloads of `AutomatedTransaction`s because `public entry` functions are immutable, so I think it would also make sense not to charge for gas when registering the payload for automation.

sjoshisupra replied:
@isaacdoidge , why do you say we should not need to validate the payload? A regular tx, if malformed, fails only once whereas the automation payload will fail at every block, that is why it is essential to validate the payload. With regards to charging for the same, why shouldn't we charge for that? We wanted to perform this validation via a native function interface, but there were technical difficulties in doing so, therefore this is being done earlier. In either case, the network is sparing resources for this validation and I do not see why they should not be charged. (CC: @aregng )

isaacdoidge replied:
We should validate the payload when the transaction is registered. After that, it cannot change and neither can the contract (at least, in a non-compatible way), so the same payload will always be valid and executable. Consequently, we should only need to validate during registration.

Aptos made the decision not to charge gas when validating the payloads of normal transactions, and I assume that they had a reason for doing this. We should remain consistent in our approach, and the easiest way to do that is to not charge for payload validation. Alternatively, we could always charge, but if we are to do that then we should make sure that we understand why Aptos initially decided not to charge.

sjoshisupra replied:
@isaacdoidge , in the comments Aptos had mentioned that charging for these things are in TODO. Also, I do not agree that just because Aptos does it in a certain fashion we should follow. Resources are consumed, so we should charge, that is my opinion.

With regards to validation, yes we do validate the payload at the time of registration. However, subsequent validation in the code is unavoidable, without some non-trivial changes, since the validation for a tx  is intertwined with unpacking/byte code verification/dependency checks and then executing.

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -875,6 +883,150 @@ impl AptosVM {
+        // Note: Feature gating is needed here because the traversal of the dependencies could
+        //       result in shallow-loading of the modules and therefore subtle changes in
+        //       the error semantics.
+        if self.gas_feature_version >= 15 {
```
This note and code block also appear in `validate_and_execute_entry_function`. Would it make sense to abstract it into a function?

aregng replied:
I haven't done much refactoring creating separate functions for duplicate code in order to make future aptos code updates easier.

isaacdoidge replied:
Won't there still be conflicts if you've modified the same file though? Anyway, fair enough. We can refactor when we have a more stable version of the AptosVM

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -875,6 +883,150 @@ impl AptosVM {
+            )
+        })?;
+
+        session.execute(|session| {
```
The code from here down in this function is duplicated in `execute_script_or_entry_function`. Is it worth abstracting it into a function?

aregng replied:
marking resolved based on this comment: https://github.com/Entropy-Foundation/aptos-core/pull/176#discussion_r1969669100

isaacdoidge commented on aptos-move/aptos-vm/src/aptos_vm_viewer.rs:
```diff
@@ -0,0 +1,88 @@
+        function: ViewFunction,
+        max_gas_amount: u64,
+    ) -> ViewFunctionOutput {
+
```
Lot's of unnecessary whitespace here.

isaacdoidge commented on aptos-move/aptos-vm/src/automated_transaction_processor.rs:
```diff
@@ -0,0 +1,480 @@
+        Ok((VMStatus::Executed, output))
+    }
+
+    fn executed_entry_function<'a, 'r, 'l>(
```
This code is very similar to `AptosVM::execute_automation_registration_txn`.

aregng replied:
marking resolved based on this comment: https://github.com/Entropy-Foundation/aptos-core/pull/176#discussion_r1969669100

isaacdoidge commented on aptos-move/aptos-vm/src/automated_transaction_processor.rs:
```diff
@@ -0,0 +1,480 @@
+
+        gas_meter.charge_intrinsic_gas_for_transaction(txn_data.transaction_size())?;
+        session.execute(|session| {
+            self.validate_and_execute_entry_function(
```
Is it necessary to re-validate the EntryFunction of an AutomatedTransaction given that we validate it when registering the task? Entry function signatures can't be modified by contract upgrades, so we don't need to do this to ensure that the registered function is still compatible with the contract. Is there something else that we need to check here that might have changed since the task was registered?

aregng replied:
This function is not only validates but also deserializes the function and arguments and prepares for execution in Move Context.
Besides at time of the registration it might be valid function, but when it is time to execute, entry function may be deleted. so it's worth to check I guess.

isaacdoidge replied:
`public entry` functions can't been modified or deleted, so it should be safe to skip verification. See https://aptos.dev/en/build/smart-contracts/book/package-upgrades.

aregng replied:
As mentioned it is not simply verification, but also deserialization and preparation for move-env.

isaacdoidge commented on types/src/unit_tests/trusted_state_test.rs:
```diff
@@ -252,6 +252,8 @@ proptest! {
     }
 
+    // Ignore for the time being until failure is investigated and fixed
+    #[ignore]
```
If these tests were passing before the feature was added then we should understand why they are now failing before we release the feature. We may not need to fix them, but we should at least know what they do and why we do or don't need to fix them.

aregng replied:
These tests where not enabled at all in CI, so when I have enabled the tests execution as I have added new tests for feature which needed to be run.
As the feature development had higher priority and I could not figure out the issue by quick investigation, they have been marked as ignored. 
Maximum I can do is to open a ticket to follow-up on it if it is ever required.

isaacdoidge replied:
Okay, if they weren't enabled before you started your work then no need to do anything further.

isaacdoidge commented on types/src/transaction/user_transaction_context.rs:
```diff
@@ -1,7 +1,51 @@
+/// EF describes entry-function type in the particular context of its reference.
+/// MSF describes multisig payload type in the particular context of its reference.
+#[derive(Debug)]
+pub enum PayloadTypeReference<EFP: Debug, MSP: Debug> {
```
Why do we need generics here? Can we not make this a non-generic enum and use `EntryFunctionPayload` and `MultisigPayload` directly?

aregng replied:
It is utilized in 2 different contexts: `UserTransactionContext` and `TransactionMetadata` and the inner value is different in the contexts, one is `EntryFunctionPayload`, the other one `EntryFunction`.

isaacdoidge replied:
Okay, thanks for clarifying.

isaacdoidge commented on types/src/transaction/automated_transaction.rs:
```diff
@@ -0,0 +1,442 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "AutomatedTransaction {{ \n \
```
Parsing logs will become much easier if you write the json serialization of the struct instead. Either that, or output the struct on a single line. Multi-line logs are much harder to parse.

isaacdoidge commented on types/src/transaction/automated_transaction.rs:
```diff
@@ -0,0 +1,442 @@
+}
+
+/// PartialEq ignores the cached OnceCell fields that may or may not be initialized.
+impl PartialEq for AutomatedTransaction {
```
Where is this implementation used? It would be helpful to add some context about how it is expected to be used in the docstring, since I notice that it doesn't factor in the block height.

aregng replied:
`Eq` requries `PartialEq`, and I just noticed that it is incomplete as block height is not included in the equality formula. I will fix it.
And it is Okay to have equality trait implemented, as it will be handy at least in tests.

isaacdoidge commented on types/src/transaction/automated_transaction.rs:
```diff
@@ -0,0 +1,442 @@
+
+    /// Sequence number of this transaction. This must match the sequence number
+    /// stored in the sender's account at the time the transaction executes.
+    pub(crate) sequence_number: Option<u64>,
```
Is this builder supposed to build `AutomatedTransaction`s, or `AutomationTransactions`? Don't only the latter (i.e. the task registration transactions) have sequence numbers?

aregng replied:
It is to build `AutomatedTransaction`. The sequence number of the `AutomatedTransaction::sequence_number` matches the automation-task index in registry.

isaacdoidge replied:
Would it make sense to rename `sequence_number` to `task_index` then? Either way, please update the docstring to communicate this.

isaacdoidge commented on aptos-move/framework/supra-framework/sources/transaction_validation.move:
```diff
@@ -167,6 +171,40 @@ module supra_framework::transaction_validation {
+    )  {
+        let gas_payer = signer::address_of(&sender);
+
+        assert!(chain_id::get() == chain_id, error::invalid_argument(PROLOGUE_EBAD_CHAIN_ID));
```
Can these checks (except the last) be abstracted out into a function? They also appear in `prologue_common`.

aregng replied:
It's doable.

isaacdoidge commented on aptos-move/framework/supra-framework/doc/aggregator_factory.md:
```diff
@@ -217,7 +217,7 @@ Returns a new aggregator.
 <td>To create a new aggregator instance, the aggregator factory must already be initialized and exist under the Supra account.</td>
 <td>High</td>
-<td>The create_aggregator_internal function asserts that AggregatorFactory exists for the Supra account.</td>
+<td>The create_aggregator_internal function asserts that AggregatorFactory exists for the Aptos account.</td>
```
There seems to be something automatically changing `Supra` back to `Aptos` in some of the docs whenever they're recompiled. It would be good to find what this is and fix it.

aregng replied:
I thought I have fixed it to be Supra, I will apply a fix one more time.

isaacdoidge commented on aptos-move/e2e-testsuite/src/tests/vm_viewer.rs:
```diff
@@ -0,0 +1,108 @@
+    let viewer_ifc_time = Instant::now();
+    let time = Instant::now();
+    let vm_viewer = AptosVMViewer::new(test_executor.data_store());
+    println!("AptosVMViewer creation time: {}", time.elapsed().as_secs_f64());
```
Would `debug` be better here? How can we turn this output off?

aregng replied:
it's a test, why `println!` can be an issue at all ?

isaacdoidge replied:
Because `println` always prints to the console, which clutters the test output. `debug` and other tracing log levels can be turned off unless debugging.

isaacdoidge commented on aptos-move/e2e-testsuite/src/tests/transaction_fuzzer.rs:
Can you please link the StackOverflow issue so we can track it?

aregng replied:
There is no issue, the test simply fails with stack-overflow error.

isaacdoidge commented on aptos-move/aptos-vm/src/transaction_validation.rs:
```diff
@@ -334,3 +447,31 @@ pub(crate) fn run_failure_epilogue(
+        features,
+        traversal_context,
+    )
+        .or_else(|e| {
```
This formatting seems unusual.

aregng replied:
noted

isaacdoidge commented on aptos-move/aptos-vm/src/errors.rs:
```diff
@@ -106,7 +110,8 @@ pub fn convert_prologue_error(
         },
         VMStatus::MoveAbort(location, code) => {
-            let new_major_status = match error_split(code) {
+            let (category, ccode) = error_split(code);
```
Why do this reassignment? The new variables aren't used anywhere else, so the original code was more concise.

aregng replied:
it;s simply left-over form some testing. Will revert to original state

isaacdoidge commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -0,0 +1,1063 @@
+        borrow_global<ActiveAutomationRegistryConfig>(@supra_framework).main_config
+    }
+
+    #[view]
```
It would be helpful to have all `view` and `entry` functions at the top of the module, since users only interact with them.

aregng replied:
noted

aregng committed:
[AN-Issue-1485] Enabled means to configure automation registry conditionally or manually (#172)


* [AN-Issue-1485] Enabled means to configure automation registry conditionally or manually

- Added a new `AutomationRegistryConfig` configuration type  enclosing
  registry configuration parameters

- Updated `GenesisConfigurations` with optional automation-registry-configuration paramter

- Introduced  function `genesis::initialize_supra_native_automation` to
  initialize auomation registry externally

- Initialized  `automation_registry` during genesis transcation
  generation conditionaly based on corresponding configuration paramters
  availability

- Introduced  feature flag which will allow conditionally fully enable
  supra native automation usage

* Guarded automation registry functions execution

- `automation_registry::on_new_epoch` will do noop if registry is not properly initialized
- `automation_registry::register` will fail task registration if `SUPRA_NATIVE_AUTOMATION` is not enabled

* Cosmetic changes

* Added view functions
 - Added view functions to query registered tasks count and ids
 - Added view functions to query registry initialization status

* Fixed test failures and removed redundant code

* Updated SUPRA_NATIVE_AUTOMATION flag value to 88

* Updated features readme

* Cosmetic changes

- make the genesis::initialize_automation_registry public to allow calls form script
- fixed couple of typos in the code and warnings

---------

Co-authored-by: Aregnaz Harutyunyan <>

nizam-supraoracles committed:
Automation/Congestion fee Implementation (#173)

* automation automation - charge `flat_registration_fee` at the time of registration which is non refundable

* automation registry - on_new_epoch fee charges for an task

* auomation fee charges - create `calculate_automation_congestion_fee` seprate function and address pr comments

* automation registry fee - create new function to handle `process_automation_task_fees` and emit event based on the execusion + few function name update

* automation registry - check if `automation_fee_cap_for_epoch` execeed then cancal this task

* automation registry - make consistance for name of task_index for entire module

* remove `refund_automation_task_fee` function as we are not refunding

* create `cleanup_and_activate_tasks` function

* remove active_task_ids, and get active task from latest list to avoid mistakes

* rename few variables and function + optimize calculation

* update automation registry events and names, also include missing fee charges event

* remove duplicate fields in `AutomationTaskFee` struct

* add sorting before start fee withdraw from automation task

* automation registry minor nitpick

* autogenerated files update

* fix cosmetic comments

* implement automation epoch change fee refund

* registry module - add upscale and downscale methods

* automation registry - add `get_automation_epoch_info` method to fetch automation epoch info

* automation registry - add new filed to handle `epoch_freeze_fees` which is not withdrawable and provied view function for the same

* rename `epoch_freeze_fees` to `epoch_locked_fees`

* store `total_committed_max_gas_amount` in the storage so the time of refund we will take it from here, + some consometic changes

* pr comments address

* Added unit tests and applied cosmentic changes

- added unit tests for automation fee charge and refund functionality

- updated fee-charge and refund functionality to take into account
  feature flag value and locked fee amount
    - Now if there is no locked fee amount no refund is attempted
    - If the feature is disabled refund. cleanup and config updates
      are done on new epoch but no task is charged automation fee for epoch as
      they will not be even scheduled for execution

- updated config update api to do check against epoch-interval and a new task
  duration cap. Now update will fail if epoch interval is greater than
  the new task-durationr- cap

- applied some cosmetic changes (formating, name-correction)

* Added view functions to estimate automation fees

* Addressed review comments and updated unit tests

* Fixed test failures

* Addressed review comments

---------

Co-authored-by: Aregnaz Harutyunyan <>

nizam-supraoracles committed:
Refactor - congestion fee increases polynomially based on exponent value as the degree (#192)

* refomat - code auto format using move plugin

* automation registry - introduce new `congestion_exponent` field in Automation Config which use for congestion fee calculation

* congestion fee calculation update as per congestion_exponent, so created `calculate_exponentiation` function and add/update unit testcases

* add `congestion_exponent` field in AutomationRegistryConfigV1 struct in rust layer

* congestion threshould should not exceed 100, add that condition

* make `congestion_exponent` type to u8

* congestion_exponent must be non-zero condition add

* `threshold_surplus_normalized + threshold_percentage_scaled` should not surplus 1, if in that case clip value of threshold_surplus

* add more comments on `calculate_exponentiation` function

* congestion_threshold_percentage is less-than equal to 100, instead less-than

Co-authored-by: Saurabh Joshi <106232965+sjoshisupra@users.noreply.github.com>

---------

Co-authored-by: Saurabh Joshi <106232965+sjoshisupra@users.noreply.github.com>

aregng committed:
[AN-Issue-1517] Removed expiry-threshold property from AutomatedTransactionBuilder (#193)

* [AN-Issue-1517] Removed expiry-threshold property from AutomatedTransactionBuilder

- Addressed review comments posted on main branch


---------

Co-authored-by: Aregnaz Harutyunyan <>

aregng committed:
[AN-Issue-1526] Checked automation fee cap for epoch against current estimated one (#194)

* [AN-Issue-1526] Checked autoamtion fee cap for epoch against current estimated one

- added non-zero check for registry_max_gas_cap configuration parameter

- alligned parameters' checks in automation_registry::initialize with
  automation_registry::update_config

* Addressed review comments

* Addressed review comments

* Cached active task indexes for epoch in registry state on new epoch change

* Updated doc files

* Avoided extra copy my moving the vector instead

---------

Co-authored-by: Aregnaz Harutyunyan <>

aregng committed:
[AN-Issue-1628] Added new automation configuration parameter task_capacity (#195)

* [AN-Issue-1628] Added new automation configuration parameter task_capacity

- Updated register function to reject a new task registration if registry is full, i.e. task_capacity has been reached

* Cosmetic changes

---------

Co-authored-by: Aregnaz Harutyunyan <>

aregng committed:
[AN-Issue-1687] Optimized fee calculation when any multiplier is 0 (#198)

Co-authored-by: Aregnaz Harutyunyan <>

aregng committed:
[AN-Issue-1688] Updated automation registry to use `coin::transfer` instead of `supra_account::transfer` (#199)

* [AN-Issue-1688] Updated automation registry to use coin::transfer instead of supra_account::transfer

* Addressed review comment

* Fixed typo

---------

Co-authored-by: Aregnaz Harutyunyan <>
